pkg/kinflate/resource/configmap.go

```diff
@@ -39,29 +44,83 @@ func NewFromConfigMap(cm manifest.ConfigMap) (*Resource, error) {
 	return &Resource{Data: data}, nil
 }
 
-func makeConfigMap(cm manifest.ConfigMap) (*corev1.ConfigMap, error) {
+func makeConfigMap(cm manifest.ConfigMap, l loader.Loader) (*corev1.ConfigMap, error) {
 	corev1cm := &corev1.ConfigMap{}
 	corev1cm.APIVersion = "v1"
 	corev1cm.Kind = "ConfigMap"
 	corev1cm.Name = cm.Name
 	corev1cm.Data = map[string]string{}
 
-	// TODO: move the configmap helpers functions in this file/package
-	if cm.EnvSource != "" {
-		if err := cutil.HandleConfigMapFromEnvFileSource(corev1cm, cm.EnvSource); err != nil {
-			return nil, err
+	envPairs, err := keyValuesFromEnvFile(l, cm.EnvSource)
+	if err != nil {
+		return nil, fmt.Errorf("error reading keys from env source file: %s %v", cm.EnvSource, err)
+	}
+
+	literalPairs, err := keyValuesFromLiteralSources(cm.LiteralSources)
+	if err != nil {
+		return nil, fmt.Errorf("error reading key values from literal sources: %v", err)
+	}
+
+	filePairs, err := keyValuesFromFileSources(l, cm.FileSources)
+	if err != nil {
+		return nil, fmt.Errorf("error reading key values from file sources: %v", err)
+	}
+
+	allPairs := append(append(envPairs, literalPairs...), filePairs...)
+
+	// merge key value pairs from all the sources
+	for _, kv := range allPairs {
+		if errs := validation.IsConfigMapKey(kv.key); len(errs) != 0 {
+			return nil, fmt.Errorf("%q is not a valid key name for a ConfigMap: %s", kv.key, strings.Join(errs, ";"))
 		}
+		if _, exists := corev1cm.Data[kv.key]; exists {
+			return nil, fmt.Errorf("cannot add key %s, another key by that name already exists: %v.", kv.key, corev1cm.Data)
+		}
+		corev1cm.Data[kv.key] = kv.value
+	}
+
+	return corev1cm, nil
+}
+
+func keyValuesFromEnvFile(l loader.Loader, path string) ([]kvPair, error) {
+	if path == "" {
```

Comment:
Do we need this check? I think loader has already handled it.

Reply:
At the expense of the callee :-)